typescript-rewrite/samlang-core/src/interpreter/expression-interpreter.ts

```diff
@@ -0,0 +1,274 @@
+          }
+          case '>': {
+            const v1 = this.eval(expression.e1) as IntValue;
+            const v2 = this.eval(expression.e2) as IntValue;
+            return { type: 'bool', value: v1.value > v2.value };
+          }
+          case '>=': {
+            const v1 = this.eval(expression.e1) as IntValue;
+            const v2 = this.eval(expression.e2) as IntValue;
+            return { type: 'bool', value: v1.value >= v2.value };
+          }
+          case '==': {
+            const v1 = this.eval(expression.e1);
+            const v2 = this.eval(expression.e2);
+            if (v1.type === 'functionValue' || v2.type === 'functionValue') {
+              throw new PanicException('Cannot compare functions!');
+            }
+            return { type: 'bool', value: isSameValue(v1, v2) };
+          }
+          case '!=': {
+            const v1 = this.eval(expression.e1);
+            const v2 = this.eval(expression.e2);
+            if (v1.type === 'functionValue' || v2.type === 'functionValue') {
+              throw new PanicException('Cannot compare functions!');
+            }
+            return { type: 'bool', value: !isSameValue(v1, v2) };
+          }
+          case '&&': {
+            const v1 = this.eval(expression.e1) as BoolValue;
+            return !v1.value ? { type: 'bool', value: false } : this.eval(expression.e2, context);
+          }
+          case '||': {
+            const v1 = this.eval(expression.e1) as BoolValue;
+            return v1.value ? { type: 'bool', value: true } : this.eval(expression.e2, context);
+          }
+          case '::': {
+            const v1 = this.eval(expression.e1) as StringValue;
+            const v2 = this.eval(expression.e2) as StringValue;
+            return { type: 'string', value: v1.value + v2.value };
+          }
+        }
+      }
+      // eslint-disable-next-line no-fallthrough
+      case 'IfElseExpression': {
+        return this.eval(
+          (this.eval(expression.boolExpression) as BoolValue).value ? expression.e1 : expression.e2,
+          context
+        );
+      }
+      case 'MatchExpression': {
+        const matchedValue = this.eval(expression.matchedExpression, context) as VariantValue;
+        const matchedPattern =
+          expression.matchingList.find((el) => el.tag === matchedValue.tag) ??
+          this.blameTypeChecker();
+        let ctx = context;
+        if (matchedPattern.dataVariable) {
+          ctx = { ...context };
+          ctx.localValues.set(matchedPattern.dataVariable, matchedValue.data);
+        }
+        return this.eval(matchedPattern.expression, ctx);
+      }
+      case 'LambdaExpression':
+        return {
+          type: 'functionValue',
+          arguments: expression.parameters.map((param) => param[0]),
+          body: expression.body,
+          context,
+        };
+      case 'StatementBlockExpression': {
+        const { block } = expression;
+        const currentContext = { ...context };
```

Comment:
I think you need a deep copy over the map content as well. It might help you figure out when it's necessary by changing the map type to `ReadonlyMap`

Reply:
Converted the map types in `interpretation-context` to `Readonly<Record<string, ClassValue | undefined>>` so with my current changes, spreading should suffice with deep copy I think